### PR TITLE
Fix #1644: Document the coding style.

### LINKED ISSUE #1644: docs: document general style for contributing
Since scala-js has some strict requirements when it comes to style, commit messages, etc, I think it would be beneficial to write down a document people can look at to make sure their contributions comply to the standard

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0 and @nicolasstucki.

Also /cc @andreaTP @yawnt, and pretty much everyone who wants to point out something I have forgotten.

File: CODINGSTYLE.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
+# Coding Style Guide
+
+The Scala.js project has pretty drastic rules in terms of coding style.
+This document tries to document them as much as possible.
+
+A *few* of these rules are checked automatically using Scalastyle, but most of them are too complex to teach to an automated tool.
+
+In some cases, [@sjrd](https://github.com/sjrd) will just say something else on a code review.
+His opinion on a direct piece of code always takes precedence over this document.
+Feel free to point out the conflict, though: the document can be improved.
```
Comment (gzm0) [inline CODINGSTYLE.md]: As funny as this sounds. This is probably the sanest thing a project can do. I'm a bit worried about scalability though...

File: CODINGSTYLE.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+def clamp(v: Int, min: Int, max: Int): Int = {
+  if (v < min) min
+  else if (v > max) max
+  else v
+}
+```
+
+Every expression that spans more than one line should be enclosed in braces.
+For example, this is wrong:
```
Comment (gzm0) [inline CODINGSTYLE.md]: You are violating this rule in the `abs` example above...

File: CODINGSTYLE.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+```scala
+// inline lambda, hence (), hence dot-notation
+list.map(x => x * 2)
+
+// long lambda, hence braces, hence infix notation
+list map { x =>
+  if (x < 5) x
+  else x * 2
+}
```
Comment (gzm0) [inline CODINGSTYLE.md]: You might want to mention that the following is ok:

```scala
list.map { x =>
  // complicated stuff
}.toMap
```

File: CODINGSTYLE.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+All public and protected methods must have an explicit result type.
+
+Procedure syntax must not be used. `: Unit =` must be used instead.
+
+Side-effect-free methods without formal parameters should be declared without `()`, unless either a) it overrides a methode defined with `()` (such as `toString()`) or b) it implements a Java method in the Java libraries.
+
+### for comprehensions
+
+For-comprehensions must use `()` if they have a single generator without `if`, such as:
```
Comment (gzm0) [inline CODINGSTYLE.md]: Even if the rhs is a pattern? I think our codebase violates this rule (especially me)...

File: CODINGSTYLE.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+```scala
+for (i <- 0 until n)
+  doStuff(i)
+```
+
+Otherwise, it must use `{}`, and there must be one generator per line.
+The body of the `for` is then automatically surrounding by braces too, even if it is a one-liner:
```
Comment (gzm0) [inline CODINGSTYLE.md]: maybe mention that ifs may or may not be on the same line.

File: CODINGSTYLE.md [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+## Usages of higher-order methods
+
+### Option and js.UndefOr
+
+Use the higher-order methods in the APIs of `Option` and `js.UndefOr` rather than doing pattern matching.
+Note particularly the `fold` method, which should be used instead of the `map`+`getOrElse` combination.
+
+### Collections
+
+Higher-order methods should be favored over loops and tail-recursive methods wherever possible.
```
Comment (gzm0) [inline CODINGSTYLE.md]: Maybe mention that "possible" includes readability. Tail-recursion should be favored over a ridiculously complex `foldLeft` IMO. (And we do that).

File: CODINGSTYLE.md [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+
+* Never put two blank lines in a row
+* (Almost) always put a blank line between two declarations in a class (exceptions are several related consecutive ScalaDocless private fields)
+* Always put blank lines around a `case` whose body spans several lines
+* Insert blank lines at will between logical blocks of statements in a method
+
+### Braces
+
+Braces should not be used for one-liner methods, nor for two-liner `if/else` methods.
+Similarly, you should not use braces for the body of a `while`/`if`/`for` that is a one-liner (but you should if it is a two-liner `if/else`).
```
Comment (gzm0) [inline CODINGSTYLE.md]: It is not clear from this, whether the following is ok:

```scala
if (foo)
  long line
else
  long line

File: CODINGSTYLE.md (hunk 4 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Can you give an example of how/where we violate that rule?

Comment (gzm0) [inline CODINGSTYLE.md]: For example this:

```scala
val newParams = for {
  p @ js.ParamDef(name, ptpe, mutable, rest) <- params
} yield {
  js.ParamDef(name, ptpe, newMutable(name.name, mutable), rest)(p.pos)
}
```

from [GenJSCode](https://github.com/scala-js/scala-js/blob/master/compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala#L712-L716).

IMO we should absolutely allow this.

Comment (sjrd) [inline CODINGSTYLE.md]: Ah, yes, indeed. Good catch!

Comment (gzm0) [inline CODINGSTYLE.md]: I think in general, if the rhs is a pattern, or the generator is complex `{}` is better:

I'd prefer:

```scala
for {
  (a, b) <- foo
} yield {
  // ...
}
```

over 

```scala
for ((a, b) <- foo) yield {
  // ..
}
```

But also, I think we should allow this to shorten the line:

```scala
for {
  foo <- myVeryLongStatementThatDoesStuff.groupBy(_.foo.bar)
} yield {
  // ...
}
```

File: CODINGSTYLE.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
+# Coding Style Guide
+
+The Scala.js project has pretty drastic rules in terms of coding style.
```
Comment (yawnt) [inline CODINGSTYLE.md]: I would try to reformulate this in a more friendly manner.. something along the lines of

The scala.js project tries to adhere to a single coding standard, to make it easier for everyone to contribute.
This document is part of an ongoing effort to document these best practices and contributions are highly encouraged to follow them.

File: CODINGSTYLE.md (hunk 1 above)
Comment (yawnt) [inline CODINGSTYLE.md]: Same as above, something like

All contributions should follow these guidelines, although you should first and foremost follow eventual suggestions that the core team might have during code review.
Feel free to point out if something they are saying conflicts with this document, we're always looking to improve it!

File: CODINGSTYLE.md (hunk 4 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK. What about simply replacing "must" by "may" in that sentence?

Comment (gzm0) [inline CODINGSTYLE.md]: Yes. Or even better: Use "may only". (We don't want the following, do we?)

```scala
for (a <- b; c <- d) yield d
```

File: CODINGSTYLE.md (hunk 8 above)
Comment (sjrd) [inline CODINGSTYLE.md]: That's an honorable effort to be kind, but saying it the way you do would be a plain lie ^^

> to make it easier for everyone to contribute

Given how contributors first react to our dedication to coding style, I hardly believe strict coding style to make it easier for everyone to contribute. Do you honestly believe that? Besides, it really is not the point; the point is to make it easier to *maintain* the codebase, which is something most contributors never do.

> contributions are highly encouraged to follow them

This also does not do justice to the fact that we, in fact, don't accept a PR until it meets the standard.

That said, I agree my sentence is probably *too* harsh. What about this:
"The Scala.js project has a strict policy regarding coding style. This document tries to document the style in use as much as possible."
?

Comment (yawnt) [inline CODINGSTYLE.md]: Well, if you say 'highly encouraged' you're pretty much saying 'are forced to respect', except you sound kinder, but if you wanna be 100% transparent, then I think it would be okay to say something like 'strict policy' as long as you give a valid reason why.
something along the lines of (feel free to change):

.... has a strict policy regarding coding style, which is why it's still very maintainable even though it numbers $n LOCs ...

File: CODINGSTYLE.md (hunk 6 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK

File: CODINGSTYLE.md (hunk 5 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK

File: CODINGSTYLE.md (hunk 4 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Indeed.

File: CODINGSTYLE.md (hunk 3 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK.

File: CODINGSTYLE.md (hunk 2 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Right. Somehow a two-liner if/else has the status of a one-liner in most cases.

Not always, though. As a counter-example, this is still invalid:
```scala
if (x > 0)
  if (x < 5) x
  else x * 2
else
  x + 4
```

File: CODINGSTYLE.md (hunk 7 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Ah, indeed. It is ok. I'll add something about that.

File: CODINGSTYLE.md (hunk 8 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK I'll try to go in that direction. Thanks for the comments!

File: CODINGSTYLE.md (hunk 1 above)
Comment (sjrd) [inline CODINGSTYLE.md]: > I'm a bit worried about scalability though...

So far, we've managed. Replacing by "the core team" as suggested by @yawnt possibly fixes the scalability issue.

File: CODINGSTYLE.md (hunk 8 above)
Comment (gzm0) [inline CODINGSTYLE.md]: I agree that giving a reason of why we do so is a very good idea. But I also agree we should remain 100% transparent.

Proposition:

The Scala.js project has a strict policy regarding coding style. It is our firm belief that this is one of the cornerstones to maintain a clean, consistent and maintainable codebase.

This document tries to document the style in use as much as possible to make it easier for everyone to contribute. (<-- this one is true, this is the reason we write the document. Internally we never needed it, but now we do, because of the contributions we get)

Comment (sjrd) [inline CODINGSTYLE.md]: Yeah, I like that proposition :)

File: CODINGSTYLE.md (hunk 1 above)
Comment (gzm0) [inline CODINGSTYLE.md]: @yawnt while I again agree that this can be formulated friendlier, I must point out that this is not *our* (for whatever definition of our) coding style, but @sjrd's. And that is a good thing. It is **much** more important to have a consistent style than everyone being happy with his/her own little preferences.

I agree that this can sound harsh or maybe even arrogant, maybe because it evokes the fear of having to deal with an unreasonable person enforcing their dictatorship. This whole thing of course only works if the person in power is the exact opposite (for which, IMHO, @sjrd easily qualifies). But how can we express that without an essay?? :)

File: CODINGSTYLE.md (hunk 8 above)
Comment (yawnt) [inline CODINGSTYLE.md]: try to make it about the users, not about what you care about, you want the project to welcome external contributions (at least I hope so?) as long as they adhere to these guidelines..
so instead of "it is our firm belief...", how about "this is one of the cornerstones that has enabled scala.js to maintain a clean...."

File: CODINGSTYLE.md (hunk 2 above)
Comment (gzm0) [inline CODINGSTYLE.md]: I think it is only ok as rhs of an assignment / definition (goes together with the line-continuation rule).

File: CODINGSTYLE.md (hunk 1 above)
Comment (sjrd) [inline CODINGSTYLE.md]: @gzm0 lol ^^

File: CODINGSTYLE.md (hunk 2 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Indeed, I believe you're right.

File: CODINGSTYLE.md (hunk 1 above)
Comment (yawnt) [inline CODINGSTYLE.md]: @sjrd wrote scala.js so, as it happens, his code style and scala.js' coincide, but scala.js supersedes a single person, it's a project with a core team and the document should be about the core team's code style IMHO

tl;dr code style is @sjrd 's, present it as scala.js core team's style to show a united front, you're all set if you want to change something in the future and at the same time you're not giving the impression of a BDFL

**Comment by andreaTP** [comment]
First of all, tnx a lot @sjrd !!!
Following my personal suggestions:

- please, start with a more polite sentence, I agree it must be a strong point but this looks definitely scaring! Something like ```If you are reading this page you wish to contribute to this work and we really appreciate it! We will be happy to drive you to a successful PR but, to avoid most of the initial interactions on the style, please, continue on this document and refer to it ....bla bla bla ``` sounds a lot better to me.
- add some note on patterns to follow in long ```extends``` and ```with``` declarations, also where to put a ```self =>``` reference is missing.
- add a note where overloaded or strictly correlated , single liner method could avoid blank line in between
- please explain a little bit more widely (i.e. with examples and so...) the Collection paragraph that looks a bit wired, also it will be appreciated a note on ```forall``` preferred to fold when is the case.

At this point it will be a good benefit to also improve the CONTRIBUTING.md file, with things like:

- work on a branch
- we will ask you to rebase your PR just before merging it (and rebasing instructions)
- ...

File: CODINGSTYLE.md (hunk 8 above)
Comment (andreaTP) [inline CODINGSTYLE.md]: :+1:

File: CODINGSTYLE.md (hunk 1 above)
Comment (andreaTP) [inline CODINGSTYLE.md]: just to be honest at a first read I definitely not appreciate that sentence too.
I perfectly understand that it is not indeed although It looks arrogant.

You can maybe say something that sounds like:
```our coding guidelines are strongly driven by @sjrd who is actually in charge to define and maintain them, please follow his direction....```

Comment (sjrd) [inline CODINGSTYLE.md]: I guess that makes sense.

**Comment by sjrd** [comment]
Updated.
I tried to address all the concerns expressed so far.

File: CODINGSTYLE.md [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+
+When calling a method declared with an empty pair of parentheses, use `()`, except if the method is Java-defined *and* does not have side-effects.
+
+### Method definition
+
+All public and protected methods must have an explicit result type.
+
+Procedure syntax must not be used. `: Unit =` must be used instead.
+
+Side-effect-free methods without formal parameters should be declared without `()`, unless either a) it overrides a methode defined with `()` (such as `toString()`) or b) it implements a Java method in the Java libraries.
```
Comment (InTheNow) [inline CODINGSTYLE.md]: typo- methode :> method

**Comment by InTheNow** [comment]
in [this style guide](https://github.com/w3c/banana-rdf/blob/series/0.8.x/CONTRIBUTING.md#style-guide) there is special mention on implicit parameters as the project is very much typeclass based. I appreciate that scala.js is _not_ typeclass based, but I wonder if they should be mentioned, if not for the sake of completeness? Just a thought...

**Comment by sjrd** [comment]
@InTheNow What's supposed to special about implicit parameters? They're just like any other parameters.

**Comment by InTheNow** [comment]
For sure, the implicit parameters are the same. The exception, perhaps, would be if you have a long list of parameters and a short list of explicits, or vice-versa - do you enforce the same style on both lists? eg:

```
lass Foo(
    val a: Int,
    val x: Int,
    val y: Int,
    val z: Int
)(implicit p: Props, x: Ops) extends Bar with Foobar {

// or

class Foo(
    val a: Int,
    val x: Int,
    val y: Int,
    val z: Int
)(implicit
    p: Props,
    o: Ops) extends Bar with Foobar {
```

File: CODINGSTYLE.md [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+    val z: Int) extends Bar with Foobar {
+
+  // declarations start here
+```
+
+As an exception, if the constructor parameters are a (long) list of "configuration" parameters, the following format should be used instead:
+
+```scala
+class Foo(
+    val x: Int,
```
Comment (InTheNow) [inline CODINGSTYLE.md]: Picky: Some people like lining up the types/names eg :

```scala
class Foo(
    val xMassTree: Int,
    val yMe:       Int,
    val z:         Int
) extends Bar with Foobar {
```

As I'm pretty sure you _don't_ want this, if you used names of varying length you would highlight your true intent.

Comment (sjrd) [inline CODINGSTYLE.md]: Good point.

File: CODINGSTYLE.md [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+
+The blank line between two consecutive declarations in a class can sometimes be omitted, if the declarations are single-line (which also means ScalaDocless) and strongly related.
+This happens pretty rarely (mostly a series of private fields).
+The rule of thumb is to always put a blank line.
+
+### Braces
+
+Braces should not be used for one-liner methods.
+Similarly, you should not use braces for the body of a `while`/`if`/`for` that is a one-liner.
+
```
Comment (InTheNow) [inline CODINGSTYLE.md]: What about a line with many `&&` s. So in the followiing, what is allowed, not allowed, preferred etc:


```scala
val myVal = this.sign == that.sign && this.numberLength == that.numberLength && this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits) 

val myVal = this.sign == that.sign && this.numberLength == that.numberLength && 
  this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)  // 2 or 4 ?

val myVal = 
  this.sign == that.sign && 
  this.numberLength == that.numberLength && 
  this.equalsArrays(that.digits) && 
  this.equalsArrays(whatever.digits)  // 2 or 4 ?

if(this.sign == that.sign && this.numberLength == that.numberLength && this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits))

if(this.sign == that.sign && this.numberLength == that.numberLength && 
   this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)){ // 2 or 4?

if(this.sign == that.sign && 
   this.numberLength == that.numberLength && 
   this.equalsArrays(that.digits) && 
   this.equalsArrays(whatever.digits)) { // 2 or 4?

val cond1 = this.sign == that.sign && this.numberLength == that.numberLength
val cond2 = this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)
val myVal = cond1 && cond2
```

File: CODINGSTYLE.md [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+The Scala.js project has a strict policy regarding coding style.
+This is one of the cornerstones that has allowed Scala.js to maintain a clean, consistent and maintainable codebase.
+
+This document tries to document the style in use as much as possible to make it easier for everyone to contribute.
+
+A *few* of these rules are checked automatically using Scalastyle, but most of them are too complex to teach to an automated tool.
+
+Our coding style is driven and "maintained" by [@sjrd](https://github.com/sjrd), the original author of Scala.js.
+Please follow his directions if they are in conflict with the present document.
+Feel free to point out the conflict, though: we are always looking to improve this document.
```
Comment (sjrd) [inline CODINGSTYLE.md]: Porting a comment by @yawnt from a commit at https://github.com/sjrd/scala-js/commit/9346a931009c50f5565cd7553903d8ca5bfdaa28#commitcomment-11666243 :

> I think it would be better to start talking about a core team instead of linking everything to a single person. so something like "follow the directions of the core team"
> the rest looks good tho, thumbs up!

Comment (sjrd) [inline CODINGSTYLE.md]: @gzm0 WDYT? I'm afraid I'm too involved to reason about this without personal interference on my own reasoning. ^^ (if you know what confirmation bias is, well, I'm in the middle of it, and even though I identified it, it seems I can't get out ...)

File: CODINGSTYLE.md (hunk 11 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Ah, yep, these probably deserve their own section. FTR:
```scala
val myVal = 
  this.sign == that.sign && 
  this.numberLength == that.numberLength && 
  this.equalsArrays(that.digits) && 
  this.equalsArrays(whatever.digits)  // 2 or 4 ?
```
is allowed with 2 spaces, and with braces around it.
```scala
if(this.sign == that.sign && this.numberLength == that.numberLength && 
   this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)){ // 2 or 4?
```
is allowed, with a space after `if`, and 4 spaces.
More than 2 lines should be avoided, and the condition extracted in a `def` before the `if` instead, with the style of the first snippet.
```scala
val cond1 = this.sign == that.sign && this.numberLength == that.numberLength
val cond2 = this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)
val myVal = cond1 && cond2
```
is allowed if `cond1` and `cond2` are meaningful and can be given a meaningful name, in which case it is preferred. Otherwise, the style of the first snippet is preferred.

File: CODINGSTYLE.md (hunk 12 above)
Comment (gzm0) [inline CODINGSTYLE.md]: I think it is a good idea to clearly point out that this is the core teams decision to not give a dictator impression. On the other hand, it would be wrong to say that everyone on the core team has the same say about coding style. How about something like this:

The Scala.js core team has decided to designate a single developer to drive and "maintain" the project's coding style. This allows us to maintain a consistent, yet flexible style. Everyone in the core team follows the style maintainer's directions. Currently @sjrd maintains the Scala.js coding style. Please follow his directions if they are in conflict with the present document. Feel free to point out the conflict, though: we are always looking to improve this document.

File: CODINGSTYLE.md [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+#### Braces in lambdas
+
+In lambdas (anonymous functions), the opening brace must be placed before the formal arguments, and not after the `=>`:
+
+```scala
+val f = { (x: Int) =>
+  body
+}
+```
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: What if the first line is too long an we need to brake the line before the `(x: Int) => ` part. Would that following be the correct approach?

```scala
val someLongIdentifierWithHighIdentation = {  
  (x: Int) => {
    body
  }
}
```

File: CODINGSTYLE.md [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+
+For short `yield` blocks, the following can be used instead:
+
+```scala
+for (x <- xs)
+  yield x * 2
+```
+
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: We might want to add the case:
```scala
for {
  i <- 0 until n
  j <- 0 until i
} yield {
  thing(i, j)
}
```

File: CODINGSTYLE.md (hunk 13 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Almost. You should still avoid the braces around the body of the lambda, so it's this:
```scala
val someLongIdentifierWithHighIdentation = {  
  (x: Int) =>
    body
}
```

File: CODINGSTYLE.md (hunk 14 above)
Comment (sjrd) [inline CODINGSTYLE.md]: OK

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -5,6 +5,13 @@
 If you haven't read it, ***read the very important notice about the Javalib
 in the [Developer documentation](./DEVELOPING.md)***.
```
Comment (nicolasstucki) [inline CONTRIBUTING.md]: W should add here the `import java.{util => ju}` and any other naming convention that we have.

**Comment by nicolasstucki** [comment]
Thats all

File: CONTRIBUTING.md (hunk 15 above)
Comment (sjrd) [inline CONTRIBUTING.md]: OK I added a section about imports. I'm not exactly sure what other well-established shortcuts we have, so I just put the one for `java.util`. If we find others we can grow the list.

**Comment by sjrd** [comment]
Updated.

**Comment by InTheNow** [comment]
@sjrd  Firstly, thanks for addressing my comments. Secondly, I think the fact that you were able to incorporate them easily, with just a few additions here and there, suggests to me that you have achieved a very nice set of standards.

Congratulations!

File: CODINGSTYLE.md [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+This is an instantiation of the rule saying that spaces can be removed around a binary operator to highlight its higher priority wrt. its neighbors.
+
+As a reminder, avoid pattern-matching on `Option` types.
+Use `fold` instead.
+
+
+## Implementing the Java lib
+
+Special rules apply to implementing classes of the JDK (typically in `javalanglib/` or `javalib/`).
+
```
Comment (gzm0) [inline CODINGSTYLE.md]: Maybe also drop a line about licensing restrictions?

File: CODINGSTYLE.md [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+    val depthOfField: Int = 3
+)(implicit ct: ClassTag[A]) extends Bar with Foobar {
+```
+
+If everything up to the first `with`, excluded, fits on one line, then the list of `with`s should be indented so as to align with `extends`, such as:
+
+```scala
+class Foo(val x: Int) extends Bar
+                         with Foobar
+                         with AnotherTrait {
```
Comment (gzm0) [inline CODINGSTYLE.md]: I would consider deprecating this. We only use it in the `VirtualFile` implementations and it does IMO not very nicely align with the style in general. We could just do:

```scala
+class Foo(val x: Int) extends Bar
    with Foobar with AnotherTrait {
```

**Comment by gzm0** [comment]
LGTM otherwise

File: CODINGSTYLE.md (hunk 16 above)
Comment (sjrd) [inline CODINGSTYLE.md]: I think we're good on that point. We already have that on the top of these two files:
https://github.com/scala-js/scala-js/blob/master/DEVELOPING.md
https://github.com/scala-js/scala-js/blob/master/CONTRIBUTING.md
Not to mention it has nothing to do with the coding style ^^

File: CODINGSTYLE.md (hunk 17 above)
Comment (sjrd) [inline CODINGSTYLE.md]: Oh we do have it in quite a lot of places.
But I agree, it does not look like the rest of the coding style. I'll change that.

File: CODINGSTYLE.md (hunk 16 above)
Comment (gzm0) [inline CODINGSTYLE.md]: Ok.

**Comment by sjrd** [comment]
Updated with the suggested changes for `with` in class declarations.

File: CODINGSTYLE.md [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+```scala
+val someLongIdentifierWithHighIdentation = {
+  (x: Int, ys: List[Traversable[String]]) =>
+    body
+}
+```
+
+### Spaces
+
+There must not be any space before the following tokens: `: , ; )`
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: I would replace `: , ; )` with `:` `,` `;` `)`

File: CODINGSTYLE.md [hunk 19]
```diff
@@ -0,0 +1,544 @@
...
+
+There must be exactly one space after the following tokens: `: , ; if for while`
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: Same as above.  `:` `,` `;` `if` `for` `while`

File: CODINGSTYLE.md [hunk 20]
```diff
@@ -0,0 +1,544 @@
...
+Procedure syntax must not be used.
+`: Unit =` must be used instead.
+
+Side-effect-free methods without formal parameters should be declared without `()`, unless either a) it overrides a method defined with `()` (such as `toString()`) or b) it implements a Java method in the Java libraries.
+
+The signature of a method is technically a single line, and hence, if it has to be broken due to line length reasons, subsequent lines should be indented 4 spaces.
+As a reminder, the line can be broken right after a `,` or a `(` (and not, for example, after `implicit` or `:`).
+You should avoid breaking the line between the last parameter and the result type; going over the 80 characters limit is preferred in that case.
+
+### for comprehensions
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: There is a slight ambiguity between the word "for" and the keyword "for". You should put `for` instead.
```
### `for` comprehensions
```

File: CODINGSTYLE.md [hunk 21]
```diff
@@ -0,0 +1,544 @@
...
+
+For-comprehensions may only use `()` if they have a single generator without `if`, such as:
```
Comment (nicolasstucki) [inline CODINGSTYLE.md]: `for` comprehensions may only use ...

**Comment by nicolasstucki** [comment]
Thats all

**Comment by sjrd** [comment]
Updated.

**Comment by nicolasstucki** [comment]
LGTM

**Comment by sjrd** [comment]
retest this please
